host.go

```diff
@@ -63,31 +70,52 @@ func newRunner(
 	errgrp, runningCtx := errgroup.WithContext(runningCtx)
 
 	runner := &gpbftRunner{
-		certStore:   cs,
-		manifest:    m,
-		ec:          ec,
-		pubsub:      ps,
-		clock:       clock.GetClock(runningCtx),
-		verifier:    verifier,
-		wal:         wal,
-		outMessages: out,
-		runningCtx:  runningCtx,
-		errgrp:      errgrp,
-		ctxCancel:   ctxCancel,
-		equivFilter: newEquivocationFilter(pID),
+		certStore:    cs,
+		manifest:     m,
+		ec:           ec,
+		pubsub:       ps,
+		clock:        clock.GetClock(runningCtx),
+		verifier:     verifier,
+		wal:          wal,
+		outMessages:  out,
+		runningCtx:   runningCtx,
+		errgrp:       errgrp,
+		ctxCancel:    ctxCancel,
+		equivFilter:  newEquivocationFilter(pID),
+		selfMessages: make(map[uint64]map[roundPhase]*gpbft.GMessage),
 	}
 
 	// create a stopped timer to facilitate alerts requested from gpbft
 	runner.alertTimer = runner.clock.Timer(0)
 	if !runner.alertTimer.Stop() {
 		<-runner.alertTimer.C
 	}
+
 	walEntries, err := wal.All()
 	if err != nil {
 		return nil, fmt.Errorf("reading WAL: %w", err)
 	}
+	var maxInstance uint64
 	for _, v := range walEntries {
 		runner.equivFilter.ProcessBroadcast(v.Message)
+		instance := v.Message.Vote.Instance
+		if runner.selfMessages[instance] == nil {
+			runner.selfMessages[instance] = make(map[roundPhase]*gpbft.GMessage)
+		}
+		// WAL is dump. To avoid relying on it returning sorted entries or making it
```

Comment:
```suggestion
		// WAL is dumb. To avoid relying on it returning sorted entries or making it
```

Reply:
Also an option to just sort by instance with `slices.SortFunc`.

Reply:
> Also an option to just sort by instance with `slices.SortFunc`.

Even then I need to iterate to find the right index. we are already traversing all the messages for the filter. At this point returns are diminishing fast on trying to polish WAL interaction.